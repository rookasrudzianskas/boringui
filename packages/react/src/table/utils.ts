import React, {ReactNode} from "react";
import {TableCollection} from "@react-types/table";

export const hasPaginationChild = (
  children: ReactNode | undefined,
  child: React.ElementType,
): {hasPagination: boolean; rowsPerPage: number} => {
  const pagination = React.Children.toArray(children).find(
    (item) => React.isValidElement(item) && item.type === child,
  );

  if (React.isValidElement(pagination)) {
    return {
      hasPagination: true,
      rowsPerPage: pagination?.props?.rowsPerPage || 0,
    };
  }

  return {hasPagination: false, rowsPerPage: 0};
};

export const isInfinityScroll = (collection: TableCollection<unknown>) => {
  const bodyProps = collection.body?.props;

  return bodyProps?.loadingState !== undefined && typeof bodyProps?.onLoadMore === "function";
};
