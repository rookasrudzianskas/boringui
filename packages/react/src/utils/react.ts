import * as React from "react";

export function getValidChildren(children: React.ReactNode) {
  return React.Children.toArray(children).filter((child) =>
    React.isValidElement(child),
  ) as React.ReactElement[];
}

export function addChildrenClass(children: React.ReactNode, className: string) {
  const validChildren = getValidChildren(children);

  return validChildren.map((child) => {
    return React.cloneElement(child, {
      className: `${child.props.className} ${className}`,
    });
  });
}
