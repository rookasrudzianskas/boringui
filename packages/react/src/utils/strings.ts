export const kebabCase = (text: string) => {
  return text
    ?.replace?.(/([a-z])([A-Z])/g, "$1-$2")
    ?.replace?.(/\s+/g, "-")
    .toLowerCase?.();
};

export const includes = (text?: string, search?: string[]) => {
  return search?.some((item) => text?.includes(item));
};
