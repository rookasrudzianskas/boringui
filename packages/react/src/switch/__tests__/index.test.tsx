import React from "react";
import {mount, shallow, ReactWrapper} from "enzyme";

import Switch from "../index";
import {nativeEvent, updateWrapper} from "../../../tests/utils";

interface IconProps {
  className?: string;
}

const Icon: React.FC<IconProps> = ({className = ""}) => <svg className={className} />;

const expectSwitchIsChecked = (wrapper: ReactWrapper) => {
  expect(wrapper.find(".nextui-switch-checked").at(0).length).not.toBe(0);
};

const expectSwitchIsUnChecked = (wrapper: ReactWrapper) => {
  expect(wrapper.find(".nextui-switch-checked").at(0).length).toBe(0);
};

const getSwitchElement = (wrapper: ReactWrapper) => {
  return wrapper.find('[role="switch"]').at(0);
};

const expectSwitchToHaveARIAChecked = (wrapper: ReactWrapper, value: boolean) => {
  expect(getSwitchElement(wrapper).props()["aria-checked"]).toBe(value);
};

const expectSwitchToHaveARIADisabled = (wrapper: ReactWrapper, value: boolean) => {
  expect(getSwitchElement(wrapper).props()["aria-disabled"]).toBe(value);
};

describe("Switch", () => {
  it("should render correctly", () => {
    const wrapper = mount(<Switch />);

    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should support square and circle", () => {
    const circle = shallow(<Switch />);

    expect(() => circle.unmount()).not.toThrow();
    const square = shallow(<Switch squared />);

    expect(() => square.unmount()).not.toThrow();
  });

  it("should render correctly with an icon", () => {
    const wrapper = mount(<Switch icon={<Icon />} />);

    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should render correctly with an iconOn", () => {
    const wrapper = mount(<Switch iconOn={<Icon />} />);

    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should render correctly with an iconOff", () => {
    const wrapper = mount(<Switch iconOff={<Icon />} />);

    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should render correctly with both icons on/off", () => {
    const wrapper = mount(<Switch iconOff={<Icon />} iconOn={<Icon />} />);

    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should trigger events when switch changed and change icons on/off", async () => {
    const wrapper = mount(
      <Switch
        iconOff={<Icon className="icon-off" />}
        iconOn={<Icon className="icon-on" />}
        initialChecked={false}
      />,
    );

    let icon = wrapper.find(".icon-off").at(0).getDOMNode() as SVGSVGElement;

    expect(icon).not.toBeNull();

    wrapper.find("input").simulate("change", {
      ...nativeEvent,
      target: {checked: true},
    });
    await updateWrapper(wrapper);
    expectSwitchIsChecked(wrapper);

    icon = wrapper.find(".icon-on").at(0).getDOMNode() as SVGSVGElement;
    expect(icon).not.toBeNull();
  });

  it("should throw error when icon is left over", () => {
    let errorMessage = "";
    const errorSpy = jest.spyOn(console, "error").mockImplementation((msg) => (errorMessage = msg));

    mount(<Switch icon={<Icon />} iconOn={<Icon />} />);
    expect(errorMessage).toContain('Remove props "icon"');
    errorSpy.mockRestore();
  });

  it("should work with different colors", () => {
    const wrapper = mount(
      <div>
        <Switch color="primary" />
        <Switch color="secondary" />
        <Switch color="success" />
        <Switch color="warning" />
        <Switch color="error" />
      </div>,
    );

    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should work with different sizes", () => {
    const wrapper = mount(
      <div>
        <Switch size="xs" />
        <Switch size="sm" />
        <Switch size="md" />
        <Switch size="lg" />
        <Switch size="xl" />
      </div>,
    );

    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should set switch follow checked prop", async () => {
    const wrapper = mount(<Switch initialChecked={true} />);

    expectSwitchIsChecked(wrapper);
    expectSwitchToHaveARIAChecked(wrapper, true);

    wrapper.setProps({checked: false});
    await updateWrapper(wrapper);
    expectSwitchIsUnChecked(wrapper);
    expectSwitchToHaveARIAChecked(wrapper, false);

    wrapper.setProps({checked: true});
    await updateWrapper(wrapper);
    expectSwitchIsChecked(wrapper);
    expectSwitchToHaveARIAChecked(wrapper, true);
  });

  it("should trigger events when switch changed", async () => {
    let checked = false;
    const changeHandler = jest.fn().mockImplementation((e) => (checked = e.target.checked));
    const wrapper = mount(<Switch onChange={changeHandler} />);

    wrapper.find("input").simulate("change", {
      ...nativeEvent,
      target: {checked: true},
    });
    await updateWrapper(wrapper);
    expectSwitchIsChecked(wrapper);
    expectSwitchToHaveARIAChecked(wrapper, true);

    expect(changeHandler).toHaveBeenCalled();
    expect(checked).toEqual(true);
  });

  it("should ignore events when switch disabled", async () => {
    const changeHandler = jest.fn();
    const wrapper = mount(<Switch disabled onChange={changeHandler} />);

    expectSwitchToHaveARIADisabled(wrapper, true);

    wrapper.find("input").simulate("change", {
      ...nativeEvent,
      target: {checked: true},
    });
    await updateWrapper(wrapper);
    expectSwitchIsUnChecked(wrapper);

    expect(changeHandler).not.toHaveBeenCalled();
  });
});
