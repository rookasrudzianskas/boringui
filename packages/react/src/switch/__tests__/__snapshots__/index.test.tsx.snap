// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render correctly 1`] = `"<label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-bwQcWP-color-default nextui-c-STltJ-JJjKI-size-md nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"></span></div></label>"`;

exports[`Switch should render correctly with an icon 1`] = `"<label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-bwQcWP-color-default nextui-c-STltJ-JJjKI-size-md nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"><svg class=\\"\\"></svg></span></div></label>"`;

exports[`Switch should render correctly with an iconOff 1`] = `"<label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-bwQcWP-color-default nextui-c-STltJ-JJjKI-size-md nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"><svg class=\\"\\"></svg></span></div></label>"`;

exports[`Switch should render correctly with an iconOn 1`] = `"<label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-bwQcWP-color-default nextui-c-STltJ-JJjKI-size-md nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"><svg class=\\"\\"></svg></span></div></label>"`;

exports[`Switch should render correctly with both icons on/off 1`] = `"<label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-bwQcWP-color-default nextui-c-STltJ-JJjKI-size-md nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"><svg class=\\"\\"></svg></span></div></label>"`;

exports[`Switch should work with different colors 1`] = `"<div><label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-bwQcWP-color-primary nextui-c-STltJ-JJjKI-size-md nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"></span></div></label><label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-cMYjfR-color-secondary nextui-c-STltJ-JJjKI-size-md nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"></span></div></label><label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-eTwznY-color-success nextui-c-STltJ-JJjKI-size-md nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"></span></div></label><label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-jTdYRZ-color-warning nextui-c-STltJ-JJjKI-size-md nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"></span></div></label><label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-dlOPhH-color-error nextui-c-STltJ-JJjKI-size-md nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"></span></div></label></div>"`;

exports[`Switch should work with different sizes 1`] = `"<div><label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-bwQcWP-color-default nextui-c-STltJ-dAsftq-size-xs nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"></span></div></label><label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-bwQcWP-color-default nextui-c-STltJ-gPEQiI-size-sm nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"></span></div></label><label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-bwQcWP-color-default nextui-c-STltJ-JJjKI-size-md nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"></span></div></label><label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-bwQcWP-color-default nextui-c-STltJ-lcjiCZ-size-lg nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"></span></div></label><label data-state=\\"unchecked\\" class=\\"nextui-c-STltJ nextui-c-STltJ-bwQcWP-color-default nextui-c-STltJ-elYIjz-size-xl nextui-c-STltJ-bgIFDi-borderWeight-normal\\"><input class=\\"nextui-c-jjKPqy nextui-c-iFHsrU nextui-c-jjKPqy-gwHlSF-animated-true nextui-switch-input\\" data-state=\\"unchecked\\" tabindex=\\"-1\\" type=\\"checkbox\\"><div aria-checked=\\"false\\" aria-disabled=\\"false\\" class=\\"nextui-c-itiWTf nextui-c-huiNHE nextui-c-itiWTf-iXJPXp-animated-true nextui-switch nextui-switch--unchecked\\" data-state=\\"unchecked\\" role=\\"switch\\" tabindex=\\"0\\"><span class=\\"nextui-c-eJPKci nextui-switch-circle\\"></span></div></label></div>"`;
